// olca-ipc/script/src/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityQuery = exports.fromDictOf = void 0;
// deno-lint-ignore-file no-explicit-any
const o = __importStar(require("./schema.js"));
function fromDictOf(type) {
    switch (type) {
        case o.RefType.Actor:
            return o.Actor.fromDict;
        case o.RefType.Currency:
            return o.Currency.fromDict;
        case o.RefType.DQSystem:
            return o.DQSystem.fromDict;
        case o.RefType.Epd:
            return o.Epd.fromDict;
        case o.RefType.Flow:
            return o.Flow.fromDict;
        case o.RefType.FlowMap:
            return o.FlowMap.fromDict;
        case o.RefType.FlowProperty:
            return o.FlowProperty.fromDict;
        case o.RefType.ImpactCategory:
            return o.ImpactCategory.fromDict;
        case o.RefType.ImpactMethod:
            return o.ImpactMethod.fromDict;
        case o.RefType.Location:
            return o.Location.fromDict;
        case o.RefType.NwSet:
            return o.NwSet.fromDict;
        case o.RefType.Parameter:
            return o.Parameter.fromDict;
        case o.RefType.Process:
            return o.Process.fromDict;
        case o.RefType.ProductSystem:
            return o.ProductSystem.fromDict;
        case o.RefType.Project:
            return o.Project.fromDict;
        case o.RefType.Result:
            return o.Result.fromDict;
        case o.RefType.SocialIndicator:
            return o.SocialIndicator.fromDict;
        case o.RefType.Source:
            return o.Source.fromDict;
        case o.RefType.Unit:
            return o.Unit.fromDict;
        case o.RefType.UnitGroup:
            return o.UnitGroup.fromDict;
    }
}
exports.fromDictOf = fromDictOf;
class EntityQuery {
    constructor(id, name) {
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: id
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: name
        });
    }
    static of(q) {
        if (!q) {
            return new EntityQuery(null, null);
        }
        if (typeof q === "string") {
            return new EntityQuery(q, null);
        }
        if (typeof q === "object") {
            return new EntityQuery(q["id"], q["name"]);
        }
        return new EntityQuery(null, null);
    }
    hasId() {
        return !!this.id;
    }
    hasName() {
        return !!this.name;
    }
    isEmpty() {
        return !this.id && !this.name;
    }
    uriName() {
        return this.hasName() ? encodeURIComponent(this.name) : "";
    }
    toDict(refType) {
        const dict = {};
        if (refType) {
            dict["@type"] = refType;
        }
        if (this.id) {
            dict["@id"] = this.id;
        }
        if (this.name) {
            dict["name"] = this.name;
        }
        return dict;
    }
}
exports.EntityQuery = EntityQuery;
